solvebio/cli/data.py

```diff
@@ -159,10 +162,63 @@ def _upload_folder(
                     )
                 )
             else:
-                remote_parent = Object.get_by_full_path(
-                    remote_folder_full_path, assert_type="folder"
-                )
-                Object.upload_file(local_file_path, remote_parent.path, vault.full_path)
+                all_files.append((local_file_path, remote_folder_full_path, vault.full_path))
+
+    # Identify which remote folders already exist
+    upload_root_path, _ = Object.validate_full_path(
+        os.path.join(base_remote_path, local_start)
+    )
+    results = GlobalSearch().filter(path__prefix=upload_root_path, type="folder")
+    remote_folders_existing = set([x.full_path for x in results])
+    all_folder_parts = set()
+    for vault, remote_folder_path in all_folders:
+        subfolders = remote_folder_path.lstrip("/").split("/")
+        parent_folder_path = ""
+        for folder in subfolders:
+            folder_full_path = os.path.join(parent_folder_path, folder)
+            if folder_full_path not in remote_folders_existing and not folder_full_path.endswith(":"):
+                all_folder_parts.add(folder_full_path)
+            parent_folder_path = folder_full_path
+
+    if not dry_run:
```

Comment:
It might be really useful to push the dry_run down into _create_folder and _create_file_job so they print exactly what is happening at that level.

Reply:
Sounds good!

Do you think the GlobalSearch should run with `dry_run`?

Reply:
Since it is a read-only step yes I would say so